extensions/csv/src/main/java/io/deephaven/csv/CsvSpecs.java

```diff
@@ -279,132 +295,256 @@ public final NewTable parse(InputStream stream) throws IOException {
+        private final ArrayBackedColumnSource<TYPE> result;
+        private final WritableColumnSource<?> reinterpreted;
+        private final ChunkWrapInvoker<TARRAY> chunkWrapInvoker;
+        private long resultSize = 0;
+
+        public MySinkBase(ArrayBackedColumnSource<TYPE> result, Class<?> interpClass,
+                          ChunkWrapInvoker<TARRAY> chunkWrapInvoker) {
+            this.result = result;
+            if (interpClass != null) {
+                reinterpreted = (WritableColumnSource<?>) result.reinterpret(interpClass);
             } else {
-                columnNames = IntStream
-                        .range(0, numColumns)
-                        .mapToObj(i -> String.format("Column%d", i + 1))
-                        .collect(Collectors.toList());
+                reinterpreted = result;
             }
+            this.chunkWrapInvoker = chunkWrapInvoker;
+        }
 
-            final NewTable.Builder table = NewTable.builder();
-            int columnIndex = 0;
-            int size = -1;
-            for (String columnName : columnNames) {
-                final Parser<?> parser = parser(columnName, columnIndex, dataRecords);
-                final Array<?> array = buildArray(getColumn(columnIndex, dataRecords), parser, dataRecords.size());
-                if (size == -1) {
-                    size = array.size();
-                }
-                table.putColumns(columnName, array);
-                ++columnIndex;
+        @Override
+        public final void write(TARRAY src, int srcOffset, long destOffset, int size) {
+            if (size == 0) {
+                return;
             }
-            return table.size(size).build();
+            final long requiredCapacity = destOffset + size;
+            reinterpreted.ensureCapacity(requiredCapacity);
+            resultSize = Math.max(resultSize, requiredCapacity);
+            try (final ChunkSink.FillFromContext context = reinterpreted.makeFillFromContext(size);
+                 final RowSequence range = RowSequenceFactory.forRange(destOffset, destOffset + size - 1)) {
+                Chunk<? extends Values> chunk = chunkWrapInvoker.apply(src, srcOffset, size);
+                reinterpreted.fillFromChunk(context, chunk, range);
+            }
+        }
+
+        private interface ChunkWrapInvoker<TARRAY> {
+            Chunk<? extends Values> apply(TARRAY data, int offset, int capacity);
+        }
+
+        public ArrayBackedColumnSource<TYPE> result() {
+            return result;
+        }
+
+        public long resultSize() {
+            return resultSize;
         }
     }
 
-    private static List<String> legalizeColumnNames(CSVRecord record) {
-        final Set<String> taken = new HashSet<>(record.size());
-        final List<String> out = new ArrayList<>(record.size());
-        for (String name : record) {
-            out.add(NameValidator.legalizeColumnName(name, (s) -> s.replaceAll("[- ]", "_"), taken));
```

Comment:
Are we doing automatic column name remapping if a given column name from a header is not legal?  Is it possible to end up with two columns with the same name in that case?  If it is, I won't ask for logic to automatically do something about it, I am just concerned about giving the user a meaningful error message if that where to happen (I assume there is a way for the user to ask for explicit mappings other than the automated ones, so with a good, explicit error message the user can fix it themselves via providing remapping of their own etc).

Reply:
This also potentially relevant in this context: https://github.com/deephaven/deephaven-core/issues/1561
The NameValidator class helps with DH column name validations I think; it has several methods to check and validate and even "legalize" (change to legal something not legal).
One example of use: https://github.com/deephaven/deephaven-core/blob/e9e8a134f54004a2efdaf710c9f1f2d4da3da500/extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTools.java#L654

Reply:
At the moment, the library does two things:
1. It will complain about duplicate column names
2. It has a pluggable validator (which in the Deephaven case defaults to `NameValidator::isValidColumnName` which decides whether a column is a valid column name

At this point, the library does not automatically "fix" column names for you. Instead it just complains about them. If the caller gets a complaint, the caller has to fix it by calling setHeaders() which replaces all the headers en masse.

Hmm, it's possible that's annoying for a user that has like 50 columns in their file.

Maybe we should add a setHeaderForColumn(3, "newname") API as well
something like that... What do you think?

Reply:
> Maybe we should add a setHeaderForColumn(3, "newname") API as well
That sounds reasonable in principle, BUT... I'd prefer we do this as a separate thing (+ticket) if it is anything more than a few lines/refactor.

Reply:
Done